### DISCUSSION TITLE
Can not only delete watermark

### DISCUSSION THREAD

**Discussion by Number18-tong** [pymupdf/PyMuPDF#2994] [Q&A]
Thanks for your remarkable work.
I try to use add_redact_annot and apply_redactions to delete watermark. but the text under the watermark also be deleted. How to solve this problem?

**Comment by JorjMcKie** [comment]
There may be no solution at all. You did not supply any details on your specific situation (example file). So I will remove the enhancement tag for the time being and move your post to "Discussions".
After the situation can be fully assessed, a potential enhancement will be re-evaluated.

**Reply by Number18-tong** [reply]
[Epgo004.pdf](https://github.com/pymupdf/PyMuPDF/files/13868004/Epgo004.pdf)
Thanks for your reply, example file as above.  Can this PDF only delete the watermark without deleting the text below it?

**Reply by JorjMcKie** [reply]
Thank you for the example file.
As suspected, this watermark cannot be removed via redaction annotations, because its logic is to eradicate everything within a rectangle.
There exist dozens of different methods to apply watermarks to a page.
Therefore there exists no universal method to remove them.

In your case, pages have different watermarks which reside in special Form XObjects.
It turns out that these objects can simply be emptied (empty PDF dictionaries) to remove the watermark appearance.
Use this little snippet. As mentioned, please note that it works in this case but will fail more often than not in cases of other watermark methods.
```python
import fitz

doc = fitz.open("test.pdf")  # input file

for xref in range(1, doc.xref_length()):  # inspect all PDF objects
    try:
        source = doc.xref_object(xref)  # PDF object source
        if not "/Watermark" in source or not "/Form" in source:
            continue  # not a watermark object
    except Exception:  # PDF contains invalid xref numbers
        continue
    doc.update_object(xref, "<<>>")  # replace by empty dictionary

doc.ez_save("no-watermark.pdf")  # save compressed
```

**Reply by Number18-tong** [reply]
Thanks a lot!!!!!!
